File: BI-TJV/semestral/rest-server/src/main/java/wrzecond/controller/TjvEmployeeController.java
package wrzecond.controller;

import org.springframework.beans.factory.annotation.Autowired;

import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;
import wrzecond.employee.TjvEmployeeCreateDTO;
import wrzecond.employee.TjvEmployeeReadDTO;
import wrzecond.employee.TjvPasswordDTO;
import wrzecond.entity.TjvEmployee;
import wrzecond.service.TjvEmployeeService;

import javax.servlet.http.HttpServletRequest;

@RestController
@TjvController.Visibility (
    findAll = TjvController.VisibilitySettings.ADMIN,
    getByID = TjvController.VisibilitySettings.ADMIN
)
@RequestMapping("/employees")
public class TjvEmployeeController extends TjvController<TjvEmployee, TjvEmployeeReadDTO, TjvEmployeeCreateDTO> {

    // === PROPERTIES ===
    private final TjvEmployeeService service;

    @Autowired
    public TjvEmployeeController(TjvEmployeeService employeeService) {
        super(employeeService, employeeService);
        this.service = employeeService;
    }

    @PatchMapping("/passChange")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    void passChange (@RequestBody TjvPasswordDTO dto, HttpServletRequest request) {
        TjvEmployee employee = authenticate(request, VisibilitySettings.LOGGED);
        try {
            service.updatePassword(employee, dto.getOldPassword(), dto.getNewPassword());
        }
        catch (DataIntegrityViolationException exception) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST);
        }
    }

    @GetMapping("/login")
    TjvEmployeeReadDTO login (HttpServletRequest request) {
        TjvEmployee employee = authenticate(request, VisibilitySettings.LOGGED);
        return service.getByID(employee.getId());
    }

    @Override
    protected String getCreateLocationPath() {
        return "/employees/%d";
    }

}

File: BI-KOT/semestral/rest-server/src/main/kotlin/wrzecond/controller/TjvEmployeeController.kt
package wrzecond.controller

import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import wrzecond.TjvEmployeeCreateDTO
import wrzecond.TjvEmployeeReadDTO
import wrzecond.TjvEmployeeUpdateDTO
import wrzecond.TjvPasswordDTO
import wrzecond.entity.TjvEmployee
import wrzecond.service.TjvEmployeeService
import javax.servlet.http.HttpServletRequest

@RestController
@RequestMapping("/employees")
@TjvController.Visibility (
    findAll = TjvController.VisibilitySettings.ADMIN,
    getByID = TjvController.VisibilitySettings.ADMIN
)
class TjvEmployeeController (override val service: TjvEmployeeService)
: TjvController<TjvEmployee, TjvEmployeeReadDTO, TjvEmployeeCreateDTO, TjvEmployeeUpdateDTO> (service, service) {

    override val createLocationPath: String
        get() = "/employees/%d"

    @PatchMapping("/passChange")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun passChange (@RequestBody dto: TjvPasswordDTO, request: HttpServletRequest)
        = authenticate(request, VisibilitySettings.LOGGED) { employee: TjvEmployee? ->
            if (employee == null) throw ResponseStatusException(HttpStatus.UNAUTHORIZED)
            service.updatePassword(employee, dto.old, dto.new)
        }

    @GetMapping("/login")
    fun login (request: HttpServletRequest)
        = authenticate(request, VisibilitySettings.LOGGED) { employee: TjvEmployee? ->
            if (employee == null) throw ResponseStatusException(HttpStatus.UNAUTHORIZED)
            service.getByID(employee.id)
        }

}